### PR TITLE
Context data capture and merge for XDM mapper

### PR DESCRIPTION
**Pull request by timkimadobe**
## Description

This feature enables capture of all context data events that are handled by the Edge Bridge extension, with the ability to merge the data dictionaries of those events to create an easy-to-use output for the XDM mapper tool. 

It is possible to merge with or without key case sensitivity (ex with case insensitive merge: product.id == Product.Id == PrOdUcT.iD), and this setting applies throughout the deep merge process; given nested dictionaries with matching keys on both sides of the merge, the inner keys will also use the same case sensitivity setting.

Please note:
- ContentView button functionality related to the new context data capture APIs is for testing purposes and will be removed before the final merge
    - May be helpful to add capability to the test app to be able to send custom JSON objects as track action/state payloads to better test context data capture APIs
- See corresponding PR in Core for updated `EventType` and `EventSource` constants: https://github.com/adobe/aepsdk-core-ios/pull/830

Request for feedback:  
- The current location of the new functionality is currently in EdgeBridge.swift; would like to get thoughts on if a helper class or some other location would be more appropriate
- in `deepMerge` L#264 in EdgeBridge.swift, currently keys within the same dictionary that are equivalent in the case insensitive compare are collapsed into one entry using case sensitive ordering, with the last entry being taken (in this case, whatever the last entry's data type is is taken without any kind of merge logic with other equivalent keys). I used case sensitive ordering in this case so that the merge would be deterministic in terms of consistent results in the final merge. Without this ordering, there is no guarantee what key value pair will make it to the final merge. However, would like to get feedback on if this approach is a good way to handle key collisions within the same dictionary when using a case insensitive compare.
ex: 
```
[
"key1": "string",
"Key1": ["key2": 123],
"KEY1": ["key3":"value"]
]

// After consolidation becomes:
["KEY1":["key3":"value"]]

// NOT (that is, no deep merge when consolidating within the same dictionary):
["KEY1":["key2": 123, "key3":"value"]]
```
- as a follow-up to the above point, the logic for merging keys into the base dictionary L#289-L#293 in the case insensitive merge has no defined order, and will result in non-deterministic results (either way for the internal case and merge case i think the behavior should be consistent)
- should we control the final JSON output's key order? Either by alphabetical, value type, etc? what about for nested structures? (adds some complexity to the final output process)

## Related Issue

See [MOB-16906](https://jira.corp.adobe.com/browse/MOB-16906) for additional details

## Motivation and Context

This is to help users of the extension quickly capture and merge context data from track action/state Events for use in the XDM mapper tool.

## How Has This Been Tested?

Tested using various context data payloads sent though trackAction API that test (see test payloads in ContentView.swift):
- the merge logic with/without case sensitive key comparisons
- value type conflicts when keys collide in merge operation (both raw value type and Optional status)
- key casing conflicts when using case insensitive merge
- deep merge for nested JSON objects

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html).
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by kevinlind** [comment]
Hi @timkimadobe , I'd suggest moving the functionality out of the EdgeBridge class to a new class. Maybe something like ContextDataCapturer?  The new class will contain all the logic to capture, merge, and report the results. EdgeBridge would contain an instance of this class but only call it for the start/stop APIs and to pass Events.

**Comment by kevinlind** [comment]
@timkimadobe You have the public APIs to start/stop the capture as static functions to class EdgeBridge. Instead, can you create an EdgeBridge+PublicAPI class which defines the public APIs? Like other public APIs in the SDK, they will dispatch events to the Event Hub which are then listened for and handled by EdgeBridge. You will need to update EventType in Core to include a new EventType.edgeBridge property and possibly add one or two EventSouce properties to represent the start/stop capture events.

File: Sources/Public/EdgeBridge+PublicAPI.swift [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+// governing permissions and limitations under the License.
+//
+
+import AEPCore
+import AEPServices
+import Foundation
+
+/// Defines the public interface for the EdgeBridge extension
+public extension EdgeBridge {
+    /// Starts context data capture session, hooking into Edge Bridge's dispatch of events.
```
Comment (kevinlind) [inline Sources/Public/EdgeBridge+PublicAPI.swift:19]: Let's add a little more here. Maybe something like:

> Starts a debug context data capture session which tracks all context data handed by the Edge Bridge extension. Once the session is stopped, a report is generated containing a merged view of the captured context data keys and values. This report can then be uploaded to Data Prep for Data Collection to map the context data to XDM.
> Note, this API is intended for use during debugging and should not be used in a production environment. The SDK log level must be configured to `LogLevel.debug` or `LogLevel.trace` for the report to be printed.

File: Sources/Public/EdgeBridge+PublicAPI.swift [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        // Dispatch event to edgebridge to signal start capture of context data
+        let event = Event(name: EdgeBridgeConstants.EventNames.EDGE_BRDIGE_START_CONTEXT_DATA_CAPTURE,
+                          type: EventType.edgeBridge,
+                          source: EventSource.startCapture,
+                          data: nil)
+
+        MobileCore.dispatch(event: event)
+    }
+    
+    /// Stops context data capture session, outputting the merge result using the case sentivity setting applied, and removing captured events from memory.
```
Comment (kevinlind) [inline Sources/Public/EdgeBridge+PublicAPI.swift:30]: Maybe change to:

> Stops a debug context data capture session and prints a report of the captured context data keys and values during the session to the device logs. The report can then be uploaded to Data Prep for Data Collection to map the context data to XDM.
> Note, this API is intended for use during debugging and should not be used in a production environment. The SDK log level must be configured to LogLevel.debug or LogLevel.trace for the report to be printed.

**Review by kevinlind — COMMENTED** [review]
Don't forget to run SwiftLint and clean up any warnings reported. Running `make lint-autocorrect` should fix most of the issues.
I also noticed the new files are members of the TestAppSwiftUI Xcode target which causes issues when building the test app. The source files should only be target members of AEPEdgeBridge.

File: Sources/EdgeBridge.swift [hunk 3]
```diff
@@ -36,6 +38,14 @@ public class EdgeBridge: NSObject, Extension {
         registerListener(type: EventType.rulesEngine,
                          source: EventSource.responseContent,
                          listener: handleRuleEngineResponse)
+
+        registerListener(type: EventType.edgeBridge,
+                         source: EventSource.startCapture,
+                         listener: handleCaptureRequest)
```
Comment (addb) [inline Sources/EdgeBridge.swift:44]: I think we can have source as `EventSource.captureContent`. And in the data we can have a key `state:start` which means start capturing and `state:stop` which means stop capturing. This way we can make the listener generic and consolidated into one.

File: Sources/EdgeBridge.swift [hunk 4]
```diff
@@ -47,6 +57,25 @@ public class EdgeBridge: NSObject, Extension {
         return true
     }
 
+    private func handleCaptureContextData(event: Event) {
+        // add to existing list
+        contextDataCapturer.addEvent(event: event)
```
Comment (addb) [inline Sources/EdgeBridge.swift:62]: nit: we can use `append(event: event)` or `add(event: event)`. Just improving the readility of the function.

File: Sources/EdgeBridgeConstants.swift [hunk 5]
```diff
@@ -18,8 +18,15 @@ enum EdgeBridgeConstants {
     static let EXTENSION_VERSION = "1.0.0-beta"
     static let LOG_TAG = FRIENDLY_NAME
 
+    enum EventDataKeys {
+        static let CONTEXT_DATA_OUTPUT_WITH_MERGE = "withMerge"
+        static let CONTEXT_DATA_MERGE_IS_CASE_SENSITIVE = "isMergeCaseSensitive"
```
Comment (addb) [inline Sources/EdgeBridgeConstants.swift:23]: nit:
```
EventDataKeys {
    ContextData {
        static let MERGE = "merge"
        static let CASE_SENSITIVE = "caseSensitive"
    }
}

```

File: Sources/Public/EdgeBridge+PublicAPI.swift (hunk 2 above)
Comment (addb) [reply Sources/Public/EdgeBridge+PublicAPI.swift:30]: Do you guys think adding an option to "writeToFile" be helpful for the user?

File: Sources/EdgeBridge.swift (hunk 4 above)
Comment (addb) [reply Sources/EdgeBridge.swift]: Yes that is fine

File: Sources/EdgeBridge.swift [hunk 6]
```diff
@@ -59,19 +55,25 @@ public class EdgeBridge: NSObject, Extension {
...
+            if captureState != EdgeBridgeConstants.EventDataKeys.ContextDataKeys.CAPTURE_STATE_STOP {
+                Log.debug(label: EdgeBridgeConstants.LOG_TAG, "Received unknown capture state: '\(String(describing: captureState))'. Defaulting to stop capture.")
+            }
             // Context data capture output options with defaults
             // Merge defaults to true
-            let withMerge: Bool = event.data?[EdgeBridgeConstants.EventDataKeys.ContextDataKeys.CONTEXT_DATA_OUTPUT_WITH_MERGE] as? Bool ?? true
+            let withMerge: Bool = event.data?[EdgeBridgeConstants.EventDataKeys.ContextDataKeys.MERGE] as? Bool ?? true
             // Case sensitive key match defaults to true
-            let isMergeCaseSensitive: Bool = event.data?[EdgeBridgeConstants.EventDataKeys.ContextDataKeys.CONTEXT_DATA_MERGE_IS_CASE_SENSITIVE] as? Bool ?? true
+            let isMergeCaseSensitive: Bool = event.data?[EdgeBridgeConstants.EventDataKeys.ContextDataKeys.CASE_SENSITIVE] as? Bool ?? true
```
Comment (addb) [inline Sources/EdgeBridge.swift:76]: nit: If you think this key needs more context we can rename it to `CASE_SENSITIVE_MERGE`

File: Sources/EdgeBridgeConstants.swift [hunk 7]
```diff
@@ -20,8 +20,11 @@ enum EdgeBridgeConstants {
 
     enum EventDataKeys {
         enum ContextDataKeys {
-            static let CONTEXT_DATA_OUTPUT_WITH_MERGE = "withMerge"
-            static let CONTEXT_DATA_MERGE_IS_CASE_SENSITIVE = "isMergeCaseSensitive"
+            static let CAPTURE_STATE = "captureState"
+            static let CAPTURE_STATE_START = "start"
+            static let CAPTURE_STATE_STOP = "stop"
```
Comment (addb) [inline Sources/EdgeBridgeConstants.swift:25]: @timkimadobe we could actually skip everything and just keep one flag again. Let's say Capture based on its value true/false we can start/stop. 

So the event with type edgebridge and source captureContent
event data would have 
[
"capture": true, 
"merge" : true
 "caseSensitiveMerge": true
 ]
 
 PS: we could simplify the keys as mentioned above.

Comment (addb) [reply Sources/EdgeBridgeConstants.swift]: Yes it's better to send on stop.

File: Sources/Public/EdgeBridge+PublicAPI.swift (hunk 2 above)
Comment (addb) [reply Sources/Public/EdgeBridge+PublicAPI.swift:33]: We can take that as an enhancement task after we close this.

File: Sources/EdgeBridge.swift [hunk 8]
```diff
@@ -47,6 +53,32 @@ public class EdgeBridge: NSObject, Extension {
...
     }
 
+    private func handleCaptureContextData(event: Event) {
+        // add to existing list
+        contextDataCapturer.addEvent(event)
+    }
+
+    private func handleCaptureRequest(_ event: Event) {
+        var captureState: Bool
+        if let capture = event.data?[EdgeBridgeConstants.EventDataKeys.ContextDataKeys.CAPTURE] as? Bool {
```
Comment (addb) [inline Sources/EdgeBridge.swift:63]: We can optimize and directly capture the value with guard let.

guard let captureState = event.data?[EdgeBridgeConstants.EventDataKeys.ContextDataKeys.CAPTURE] as? Bool else {
Log.debug(label: EdgeBridgeConstants.LOG_TAG, "Received unknown capture state, ignoring this event.")
return 
}

Comment (addb) [reply Sources/EdgeBridge.swift:63]: So we will ignore the invalid values. After this we can add logic for captureState true/false as you have.

File: Sources/ContextDataCapturer.swift [hunk 9]
```diff
@@ -0,0 +1,322 @@
...
+        self.isCaseSensitive = isCaseSensitive
+        self.ids = [event.id]
+        self.dictionary = event.data ?? [:]
+        self.keySet = keySetResult
+    }
+}
+
+class ContextDataCapturer {
+
+    var contextDataCaptureIsActive: Bool = false
```
Comment (addb) [inline Sources/ContextDataCapturer.swift:72]: let's rename this to `isActive`

File: Sources/ContextDataCapturer.swift [hunk 10]
```diff
@@ -0,0 +1,322 @@
...
+    var contextDataStore: [Event] = []
+
+    /// Starts a debug context data capture session which tracks all context data handed by the Edge Bridge extension.
+    /// Once the session is stopped, a report is generated containing a merged view of the captured context data keys
+    /// and values. This report can then be uploaded to Data Prep for Data Collection to map the context data to XDM.
+    ///
+    /// Note, this API is intended for use during debugging and should not be used in a production environment. The
+    /// SDK log level must be configured to `LogLevel.debug` or `LogLevel.trace` for the report to be printed.
+    public func startCapture() {
+        contextDataCaptureIsActive = true
```
Comment (addb) [inline Sources/ContextDataCapturer.swift:82]: We need to add validation to ignore duplicate startCapture events.

```
if !isActive {
   isActive = true
    Log.T("Context Data capturer is already active and capturing.")
}
else {

 Log.T("Context Data capturer is active")
}
```

File: Sources/ContextDataCapturer.swift [hunk 11]
```diff
@@ -0,0 +1,322 @@
...
+    /// data to XDM.
+    ///
+    /// Note, this API is intended for use during debugging and should not be used in a production environment. The SDK log
+    /// level must be configured to `LogLevel.debug` or `LogLevel.trace` for the report to be printed.
+    ///
+    /// - Parameters:
+    ///     - withMerge: Controls if merge logic is applied to captured `Event`s
+    ///     - isMergeCaseSensitive: Controls if merge logic for matching keys uses case sensitive compare or not
+    public func stopCapture(withMerge: Bool, isMergeCaseSensitive: Bool) {
+        contextDataCaptureIsActive = false
```
Comment (addb) [inline Sources/ContextDataCapturer.swift:104]:
```
if !isActive {
    Log.T("Cannot stop ContextDataCapturer, as it is not active currently.")
}
else {
   outputCapturedContextData(withMerge: withMerge, isMergeCaseSensitive: isMergeCaseSensitive)
   contextDataStore.removeAll()
   Log.T("Context Data capturer has been stopped")
  isActive = false
}
```

**Comment by addb** [comment]
> Thanks for the followup review Arjun! I liked the suggested simplifications to the constants, and I've updated the logical flow where they are used to reflect the type change

Thanks for the updates @timkimadobe, I added few more comments. Feel free to change the wordings for the log messages as suited. I have written it in a way to give you an idea. Let me know if you have any doubts/concerns.

File: Sources/ContextDataCapturer.swift [hunk 12]
```diff
@@ -0,0 +1,322 @@
...
+            var mergeResult = MergeResult(event: firstEvent, isCaseSensitive: isMergeCaseSensitive, keySetResult: extractKeySet(event: firstEvent, isCaseSensitive: isMergeCaseSensitive))
+            // Start merging next events into result
+            if contextDataStore.count > 1 {
+                for i in 1..<contextDataStore.count {
+                    mergeResult = mergeEvents(mergeResult: mergeResult, eventToMerge: contextDataStore[i])
+                }
+            }
+            addToMergeReport(text: "============== Merge Result =================")
+            addToMergeReport(text: "-------------- Value type/key merge conflicts --------------")
+            for (key, value) in mergeResult.keySet {
```
Comment (addb) [inline Sources/ContextDataCapturer.swift:135]: Can we add this logic in a separate `function getMergeConflictReport(mergeResult)`

Comment (addb) [reply Sources/ContextDataCapturer.swift:145]: Yes I was thinking to move out the type/key merge conflicts logic.

File: Sources/ContextDataCapturer.swift [hunk 13]
```diff
@@ -47,122 +47,78 @@ private struct KeySet {
...
 /// metadata from merge process and merge settings.
 /// Initialized with an `Event` which will become the base of the overall merge
-private struct MergeResult {
-    /// The merge setting for whether or not key comparisons should be performed using case sensitivity or not
-    let isCaseSensitive: Bool
-    /// The array of `Event` IDs appended in merge order that were used to arrive at the final merge result
-    var ids: [UUID]
+private class CaptureResult {
+    /// The array of `Event`s appended in merge order that were used to arrive at the final merge result
+    private var events: [Event] = []
```
Comment (addb) [inline Sources/ContextDataCapturer.swift:52]: We can skip initialization here if we do it in init.

**Comment by addb** [comment]
I believe the responsibility of collecting, storing and clearing the events should be with the capturer and we should pass the entire array of events for generating the report.

getCaptureReport(events: [Event], merge: Bool, caseSensitiveMerge: Bool) -> String

**Pull request merged**